subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -542,6 +542,110 @@ static int engine_add_observer(struct lwm2m_message *msg,
 	return 0;
 }
 
+int lwm2m_engine_update_observer_min_period(char *pathstr, uint32_t period_s)
+{
+	int i, ret;
+	struct lwm2m_obj_path path;
+
+	ret = string_to_path(pathstr, &path, '/');
+	if (ret < 0) {
+		return ret;
+	}
+
+	switch (path.level) {
```

Comment:
I think this implementation is a bit lengthy unnecessarily, the lookup could be handled with single loop:

```
	for (i = 0; i < CONFIG_LWM2M_ENGINE_MAX_OBSERVER; i++) {
		if (observe_node_data[i].path.level == path.level &&
		    observe_node_data[i].path.obj_id == path.obj_id &&
		    (path.level >= 2 ?
		     observe_node_data[i].path.obj_inst_id == path.obj_inst_id : true) &&
		    (path.level >= 3 ?
		     observe_node_data[i].path.res_id == path.res_id : true))

			observe_node_data[i].min_period_sec = period_s;
			return 0;
		}
	}
```

Note, there's no need to verify `path.res_inst_id` as according to the LwM2M spec observation of individual resource instances is not allowed.

Reply:
It's indeed way better like this. Tested for my sample case and working as intended.